### PR TITLE
Onnx validate

### PR DESCRIPTION
**Pull request by II245**

### CODE REVIEW

File: models.py [hunk 1]
```diff
@@ -284,7 +285,8 @@ def forward(
 	) -> shaping.BCt:
 
 		if self.frontend is not None:
-			mask = temporal_mask(x, compute_output_lengths(x, xlen)) if xlen is not None else None
+			ol = compute_output_lengths(x, xlen)
+			mask = temporal_mask(x.squeeze(1), ol) if xlen is not None else None
```
Comment (vadimkantorov) [inline models.py]: Please eliminate duplicate `x.squeeze(1)` by putting it in a variable, e.g. `x = x.squeeze(1)` if x is not used below

File: models.py [hunk 2]
```diff
@@ -1401,3 +1403,36 @@ def __init__(self, *args, **kwargs):
...
+
+
+class OnnxModelWraper(nn.Module):
+	def __init__(self, onnx_model_path, provider=None):
+		self.dict = dict
+		self.bpe_only = False
+		super(OnnxModelWraper, self).__init__()
+		onnxruntime_session = onnxruntime.InferenceSession(onnx_model_path)
+
+		if provider:
```
Comment (vadimkantorov) [inline models.py]: `provider` -> `providers`

File: models.py [hunk 3]
```diff
@@ -1401,3 +1403,36 @@ def __init__(self, *args, **kwargs):
 			nonlinearity = ('leaky_relu', 0.01),
 			**kwargs
 		)
+
+
+class OnnxModelWraper(nn.Module):
+	def __init__(self, onnx_model_path, provider=None):
```
Comment (vadimkantorov) [inline models.py]: Please put in `['CUDAExecutionProvider', 'CPUExecutionProvider']` as the default value

File: models.py [hunk 4]
```diff
@@ -1401,3 +1403,36 @@ def __init__(self, *args, **kwargs):
...
+		self.dict = dict
+		self.bpe_only = False
+		super(OnnxModelWraper, self).__init__()
+		onnxruntime_session = onnxruntime.InferenceSession(onnx_model_path)
+
+		if provider:
+			onnxruntime_session.set_providers(list(provider))
+
+		self.model = lambda x: onnxruntime_session.run(None, dict(x=x))
```
Comment (vadimkantorov) [inline models.py]: please get rid of lambdas, they are unnecessary

File: models.py [hunk 5]
```diff
@@ -1401,3 +1403,36 @@ def __init__(self, *args, **kwargs):
...
+
+	def forward(self, x, xlen, y, ylen):
+		load_batch = lambda x: x.cpu().numpy()
+		logits = self.model(load_batch(x.squeeze(1)))
+
+		print('argmax: ', torch.from_numpy((logits[0])).argmax(dim=1))
+		logits = [torch.from_numpy(l).to(device=y.device) for l in logits]
```
Comment (vadimkantorov) [inline models.py]: please fuse `torch.from_numpy(...)` and `to(...)` into a single `torch.as_tensor(...)` call

File: models.py [hunk 6]
```diff
@@ -1401,3 +1403,36 @@ def __init__(self, *args, **kwargs):
...
+		super(OnnxModelWraper, self).__init__()
+		onnxruntime_session = onnxruntime.InferenceSession(onnx_model_path)
+
+		if provider:
+			onnxruntime_session.set_providers(list(provider))
+
+		self.model = lambda x: onnxruntime_session.run(None, dict(x=x))
+
+	def forward(self, x, xlen, y, ylen):
+		load_batch = lambda x: x.cpu().numpy()
```
Comment (vadimkantorov) [inline models.py]: please get rid of the lambda

File: models.py [hunk 7]
```diff
@@ -1401,3 +1403,36 @@ def __init__(self, *args, **kwargs):
...
+		self.bpe_only = False
+		super(OnnxModelWraper, self).__init__()
+		onnxruntime_session = onnxruntime.InferenceSession(onnx_model_path)
+
+		if provider:
+			onnxruntime_session.set_providers(list(provider))
+
+		self.model = lambda x: onnxruntime_session.run(None, dict(x=x))
+
+	def forward(self, x, xlen, y, ylen):
```
Comment (vadimkantorov) [inline models.py]: please copy over shaping type hints from the JasperNet definition, this will help you understand what's  going on

File: models.py (hunk 3 above)
Comment (vadimkantorov) [inline models.py]: Please take in logging severity as an argument, just like done in `train.py`

File: train.py [hunk 8]
```diff
@@ -125,7 +125,10 @@ def apply_model(data_loader, model, generator, text_pipelines, device, oom_handl
 		x, xlen, y, ylen = x.to(device, non_blocking = True), xlen.to(device, non_blocking = True), y.to(device, non_blocking = True), ylen.to(device, non_blocking = True)
 		with torch.no_grad():
 			try:
-				logits, log_probs, olen, loss = map(model(x, xlen, y = y, ylen = ylen).get, ['logits', 'log_probs', 'olen', 'loss'])
+				pred = model(x, xlen, y = y, ylen = ylen)
```
Comment (vadimkantorov) [inline train.py]: Please use the OnnxModelWrapper in the ONNX export branch, i.e. the output from OnnxModelWrapper should be used for the allclose check

File: models.py [hunk 9]
```diff
@@ -1401,3 +1403,57 @@ def __init__(self, *args, **kwargs):
 			nonlinearity = ('leaky_relu', 0.01),
 			**kwargs
 		)
+
+
+class OnnxModelWraper(nn.Module):
```
Comment (vadimkantorov) [inline models.py]: typo: `Wraper` -> `Wrapper`

File: scripts/validate_onnx_torch_model.sh [hunk 10]
```diff
@@ -0,0 +1,10 @@
+set -e
+
+python3 train.py $@ \
+  --device cuda \
+  --fp16 O2 \
```
Comment (vadimkantorov) [inline scripts/validate_onnx_torch_model.sh]: please start with fp32 experiments first

File: train.py [hunk 11]
```diff
@@ -481,12 +486,32 @@ def main(args):
 			onnxruntime.set_default_logger_severity(0)
 		(logits_, ) = onnxruntime_session.run(None, dict(x = waveform_input.cpu().numpy()))
 
+		wraper = models.OnnxModelWraper(args.onnx)
+		wraper_logits = wraper(waveform_input)['logits'][0]
+		args.fp16 = None
```
Comment (vadimkantorov) [inline train.py]: this is probably the cause of the problems; delete this line? setting this was enforcing tighter tolerance in the `allclose` check. the model was still running in fp16, but the `rtol` was checked as if it was fp32

File: models.py (hunk 1 above)
Comment (II245) [reply models.py]: done

File: models.py (hunk 5 above)
Comment (II245) [reply models.py]: done

File: models.py (hunk 3 above)
Comment (II245) [reply models.py]: done

File: models.py (hunk 7 above)
Comment (II245) [reply models.py]: done

File: models.py (hunk 3 above)
Comment (II245) [reply models.py]: Please give me a link to train.py example, I can't find example.

File: train.py (hunk 8 above)
Comment (II245) [reply train.py]: done

File: models.py (hunk 9 above)
Comment (II245) [reply models.py]: done

File: scripts/validate_onnx_torch_model.sh (hunk 10 above)
Comment (II245) [reply scripts/validate_onnx_torch_model.sh]: done

File: train.py (hunk 11 above)
Comment (II245) [reply train.py]: done

File: train.py [hunk 12]
```diff
@@ -120,12 +120,18 @@ def weight_stats(self, iteration, model, log_weight_distribution, eps = 1e-9):
...
 
-def apply_model(data_loader, model, generator, text_pipelines, device, oom_handler):
+def apply_model(data_loader, model, generator, text_pipelines, device, oom_handler, onnx_wrapper=None, forward_x_only=False):
 	for meta, s, x, xlen, y, ylen in data_loader:
 		x, xlen, y, ylen = x.to(device, non_blocking = True), xlen.to(device, non_blocking = True), y.to(device, non_blocking = True), ylen.to(device, non_blocking = True)
 		with torch.no_grad():
 			try:
-				logits, log_probs, olen, loss = map(model(x, xlen, y = y, ylen = ylen).get, ['logits', 'log_probs', 'olen', 'loss'])
+				model_to_infer = onnx_wrapper if onnx_wrapper is not None else model
+				infer_kwargs = {'x': x} if forward_x_only else {'x': x, 'xlen': xlen, 'y': y, 'ylen': ylen}
```
Comment (vadimkantorov) [inline train.py]: please use `dict(x = x)` and `dict(x = x, xlen = xlen, y = y, ylen = ylen)`. this is the dict syntax we're useing in all other places

File: models.py [hunk 13]
```diff
@@ -1401,3 +1404,40 @@ def __init__(self, *args, **kwargs):
...
+
+
+class OnnxWrapper():
+	def __init__(self, onnx_model_path, providers=['CUDAExecutionProvider', 'CPUExecutionProvider']):
+		self.dict = dict
+		self.bpe_only = False
+		self.onnxruntime_session = onnxruntime.InferenceSession(onnx_model_path)
+
+		if providers:
+			self.onnxruntime_session.set_providers(list(providers))
```
Comment (vadimkantorov) [inline models.py]: is `list` call necessary? can't you use `providers` directly?

File: models.py [hunk 14]
```diff
@@ -1401,3 +1404,40 @@ def __init__(self, *args, **kwargs):
 			nonlinearity = ('leaky_relu', 0.01),
 			**kwargs
 		)
+
+
+class OnnxWrapper():
```
Comment (vadimkantorov) [inline models.py]: why are you not deriving from `nn.Module`? were there problem with this?

File: models.py [hunk 15]
```diff
@@ -284,8 +285,10 @@ def forward(
 	) -> shaping.BCt:
 
 		if self.frontend is not None:
-			mask = temporal_mask(x, compute_output_lengths(x, xlen)) if xlen is not None else None
-			x = self.frontend(x.squeeze(1), mask = mask)
+			ol = compute_output_lengths(x, xlen)
```
Comment (vadimkantorov) [inline models.py]: for simplicity, can this line go after the squeeze? so that it's clear that it does not depend on being called on not-squeezed input

File: train.py [hunk 16]
```diff
@@ -120,12 +120,18 @@ def weight_stats(self, iteration, model, log_weight_distribution, eps = 1e-9):
 				self.summary_writer.add_histogram(tag + '/grad', param.grad, iteration)
 	
 
-def apply_model(data_loader, model, generator, text_pipelines, device, oom_handler):
+def apply_model(data_loader, model, generator, text_pipelines, device, oom_handler, onnx_wrapper=None, forward_x_only=False):
```
Comment (vadimkantorov) [inline train.py]: the final metrics seem to match, so no need to pass through onnx_wrapper, you can just assign `model` to `OnnxWrapper` instance

File: train.py [hunk 17]
```diff
@@ -120,12 +120,18 @@ def weight_stats(self, iteration, model, log_weight_distribution, eps = 1e-9):
...
 	for meta, s, x, xlen, y, ylen in data_loader:
 		x, xlen, y, ylen = x.to(device, non_blocking = True), xlen.to(device, non_blocking = True), y.to(device, non_blocking = True), ylen.to(device, non_blocking = True)
 		with torch.no_grad():
 			try:
-				logits, log_probs, olen, loss = map(model(x, xlen, y = y, ylen = ylen).get, ['logits', 'log_probs', 'olen', 'loss'])
+				model_to_infer = onnx_wrapper if onnx_wrapper is not None else model
+				infer_kwargs = {'x': x} if forward_x_only else {'x': x, 'xlen': xlen, 'y': y, 'ylen': ylen}
+				logits, log_probs, olen, loss = map(model_to_infer(**infer_kwargs).get, ['logits', 'log_probs', 'olen', 'loss'])
+
+				if loss is None:
```
Comment (vadimkantorov) [inline train.py:128]: please move this hack into `OnnxWrapper`. please check what should be the correct `loss` tensor's shape and put a correctly-shaped dummy

File: train.py [hunk 18]
```diff
@@ -456,7 +464,11 @@ def main(args):
 		if args.fp16:
 			model = models.InputOutputTypeCast(model.to(torch.float16), dtype = torch.float16)
 
-		waveform_input = torch.rand(args.onnx_sample_batch_size, args.onnx_sample_time, device = args.device)
+		if args.onnx_waveform_input:
+			waveform_input = torch.load(args.onnx_waveform_input).to(device = args.device).squeeze(1)
```
Comment (vadimkantorov) [inline train.py]: Please replace by `torch.load(args.onnx_waveform_input, map_location = args.device)`

Comment (vadimkantorov) [inline train.py]: why is `squeeze` needed?

File: train.py [hunk 19]
```diff
@@ -473,6 +485,7 @@ def main(args):
 				0: 'B', 2: 't'
 			})
 		)
+
 		onnxruntime_session = onnxruntime.InferenceSession(args.onnx)
```
Comment (vadimkantorov) [inline train.py]: please replace this code by using the wrapper

File: train.py [hunk 20]
```diff
@@ -481,12 +494,37 @@ def main(args):
 			onnxruntime.set_default_logger_severity(0)
```
Comment (vadimkantorov) [inline train.py]: Please support passing severity in the Wrapper's constructor

File: train.py [hunk 21]
```diff
@@ -481,12 +494,37 @@ def main(args):
...
+
+		print('pytorch with wrapper max difference: ', (logits - wraper_logits).abs().max())
+
+		pytorch_with_onnx = torch.allclose(
 				logits.cpu(),
 				torch.from_numpy(logits_),
 				**{'rtol': 1e-01, 'atol': 1e-02} if args.fp16 else {'rtol': 1e-02, 'atol': 1e-03}
 		)
 
+		wrapper_with_onnx = torch.allclose(
```
Comment (vadimkantorov) [inline train.py]: I guess all this code is not needed anymore

File: models.py (hunk 3 above)
Comment (vadimkantorov) [reply models.py]: GitHub allows to search inside the repo. It doesn't always work, but in this case it does: https://github.com/vadimkantorov/convasr/search?q=severity

File: train.py [hunk 22]
```diff
@@ -481,12 +494,37 @@ def main(args):
 			onnxruntime.set_default_logger_severity(0)
 		(logits_, ) = onnxruntime_session.run(None, dict(x = waveform_input.cpu().numpy()))
 
-		assert torch.allclose(
+		wraper = models.OnnxWrapper(args.onnx)
```
Comment (vadimkantorov) [inline train.py]: when you're fixing a typo, please do the search everywhere. this typo was already fixed in the class name

File: train.py [hunk 23]
```diff
@@ -481,12 +494,37 @@ def main(args):
...
+		wraper_logits = wraper(waveform_input)['logits'][0]
```
Comment (vadimkantorov) [inline train.py]: same as above

Comment (II245) [reply train.py]: done

File: train.py (hunk 22 above)
Comment (II245) [reply train.py]: done

File: train.py (hunk 21 above)
Comment (II245) [reply train.py]: done

File: train.py (hunk 20 above)
Comment (II245) [reply train.py]: done

File: train.py (hunk 19 above)
Comment (II245) [reply train.py]: done

File: train.py (hunk 18 above)
Comment (II245) [reply train.py]: The waveform was exported with shape tree dimensions. Should I check the frontend in model params and assert this case?

Comment (II245) [reply train.py]: done

File: train.py (hunk 17 above)
Comment (II245) [reply train.py:128]: done

File: train.py (hunk 16 above)
Comment (II245) [reply train.py]: done

File: models.py (hunk 15 above)
Comment (II245) [reply models.py]: done

File: models.py (hunk 14 above)
Comment (II245) [reply models.py]: Now nested from nn.Module

File: models.py (hunk 13 above)
Comment (II245) [reply models.py]: done

File: train.py (hunk 12 above)
Comment (II245) [reply train.py]: done

File: train.py (hunk 20 above)
Comment (vadimkantorov) [reply train.py]: Have you tested if severity will get set with `--verbose`? I think it currently has a bug

File: train.py (hunk 18 above)
Comment (vadimkantorov) [reply train.py]: please add a comment in the code about this

File: train.py [hunk 24]
```diff
@@ -1006,6 +1014,9 @@ def init_process(args):
 		'--window', default = 'hann_window', choices = ['hann_window', 'hamming_window'], help = 'for frontend'
 	)
 	parser.add_argument('--onnx')
+	parser.add_argument('--validate-onnx', action = 'store_true', help = 'flag is need to set if you validate onnx model with OnnxWrapper without export from pytorch',)
```
Comment (vadimkantorov) [inline train.py]: please follow the convention, rename it into '--onnx-validate'

File: train.py (hunk 18 above)
Comment (II245) [reply train.py]: done

File: train.py (hunk 20 above)
Comment (II245) [reply train.py]: Fix a bug.

File: train.py (hunk 24 above)
Comment (II245) [reply train.py]: done
